msu/classes/event.nut

```diff
@@ -0,0 +1,72 @@
+::MSU.Class.Event <- class
+{
+	__Functions = null;
+
+	constructor()
+	{
+		this.__Functions = [];
+	}
+
+	function register( _env, _function )
+	{
+		// ::MSU.requireFunction(_function);
+		foreach (i, funcInfo in this.__Functions)
+		{
+			if (funcInfo[1] == _function)
+			{
+				return;
+			}
+		}
+
+		this.__Functions.push([_env, _function]);
+	}
+
+	function unregister( _function )
+	{
+		// ::MSU.requireFunction(_function);
+		foreach (i, funcInfo in this.__Functions)
+		{
+			if (funcInfo[1] == _function)
+			{
+				this.__Functions.remove[i];
+				return;
+			}
+		}
+
+		throw ::MSU.Exception.KeyNotFound(_function);
+	}
+
+	function invoke( _argsArray = null )
+	{
+		if (_argsArray == null) _argsArray = [null];
+		else _argsArray.insert(0, null);
+
+		foreach (funcInfo in this.__Functions)
+		{
+			_argsArray[0] = funcInfo[0];
+			funcInfo[1].acall(_argsArray);
+		}
+	}
+
+	// Will iterate over every registered function and check _conditionFunction
+	// if _conditionFunction returns false, stops iteration immediately
```

Comment:
Feels to me like this should perhaps be based on the actual functions that are being called,ie if they return `false` then stop or something along those lines.

Reply:
Nah, those functions can be anything. They don't necessarily have to be Boolean returning functions. This `conditionalInvoke` is more along the lines of replicating the `isAlive` feature of skill_container events where the iteration stops if the actor is no longer alive.

Reply:
Returning false to stop an event from continuing feels quite natural, it's how it's done in JS event bubbling for example. Using some otherwise decoupled thing to check for a state is weird in contrast.